File: src/maze.cc
#include "maze.h"

#include <algorithm>
#include <chrono>
#include <iostream>
#include <thread>
#include <queue>
#include <unordered_set>

#include "rang.h"
#include "utils.h"

Maze::Maze(string maze_str) {
    auto lines = Utils::split_str(maze_str, "\n");

    for (int i = 0; i < lines.size(); i++) {
        vector<bool> row;
        auto chars = Utils::split_str(lines[i], "");

        for (int j = 0; j < chars.size(); j++) {
            if (chars[j] == "A") {
                this->start = make_tuple(i, j);
            } else if (chars[j] == "B") {
                this->goal = make_tuple(i, j);
            }
            row.push_back(chars[j] == "#");
        }
        this->walls.push_back(row);
    }

    this->dimensions = make_tuple(lines.size(), lines[0].size());
}

void Maze::solve(function<int(Position)> heuristic) {
    auto cmp = [](shared_ptr<Node> left, shared_ptr<Node> right) { return left->f > right->f; };
    priority_queue<shared_ptr<Node>, vector<shared_ptr<Node>>, decltype(cmp)> frontier(cmp);
    unordered_set<Position> explored;

    shared_ptr<Node> start_node = make_shared<Node>(Node{this->start});
    start_node->h = heuristic ? heuristic(this->start) : manhattan_distance(this->start, this->goal);
    start_node->f = start_node->g + start_node->h;
    frontier.push(start_node);

    while (!frontier.empty()) {
        shared_ptr<Node> node = frontier.top();
        frontier.pop();

        if (node->position == this->goal) {
            this->solution = node;
            return;
        }

        explored.insert(node->position);
        this->explored_states.push_back(node);

        auto moves = this->possible_moves(node->position);
        for (Position pos : moves) {
            if (explored.find(pos) != explored.end()) {
                continue;
            }

            shared_ptr<Node> neighbour = make_shared<Node>(Node{pos, node});
            neighbour->g = node->g + 1;
            neighbour->h = heuristic ? heuristic(pos) : manhattan_distance(pos, this->goal);
            neighbour->f = neighbour->g + neighbour->h;

            auto is_in_frontier = any_of(frontier.c.begin(), frontier.c.end(), [&neighbour](shared_ptr<Node> n) {
                return n->position == neighbour->position;
            });

            if (!is_in_frontier) {
                frontier.push(neighbour);
            }
        }
    }
}

int Maze::a_star_heuristic(Position pos) {
    return manhattan_distance(pos, this->goal);
}

void Maze::print_solution() {
    vector<Position> sol_pos;
    if (this->solution != nullptr) {
        auto node = this->solution;
        while (node->parent != nullptr) {
            sol_pos.push_back(node->position);
            node = node->parent;
        }
    }

    vector<Position> expl_pos;
    for (auto node : this->explored_states) {
        expl_pos.push_back(node->position);
    }

    int count = 0;
    while (count < expl_pos.size()) {
        this->print_maze(vector<Position>(),
                         vector<Position>(expl_pos.begin(), expl_pos.begin() + count));
        this_thread::sleep_for(chrono::milliseconds(100));
        Utils::move_up_by(get<0>(this->dimensions));
        count++;
    }

    if (this->solution != nullptr) {
        count = 0;
        while (count < sol_pos.size()) {
            this->print_maze(vector<Position>(sol_pos.begin(), sol_pos.begin() + count), expl_pos);
            this_thread::sleep_for(chrono::milliseconds(50));
            Utils::move_up_by(get<0>(this->dimensions));
            count++;
        }
    }

    this->print_maze(sol_pos, expl_pos);
    cout << endl;

    if (this->solution != nullptr)
        cout << "Path length:     " << sol_pos.size() << endl;
    else
        cout << "No solution found" << endl;
    cout << "Explored states: " << this->explored_states.size() << endl;
}

void Maze::print_maze(vector<Position> solution_pos, vector<Position> explored_pos) {
    for (int i = 0; i < this->walls.size(); i++) {
        for (int j = 0; j < this->walls[0].size(); j++) {
            Position pos = make_tuple(i, j);
            auto is_sol_pos = any_of(solution_pos.begin(), solution_pos.end(),
                                     [&pos](Position p) { return p == pos; });
            auto is_expl_pos = any_of(explored_pos.begin(), explored_pos.end(),
                                      [&pos](Position p) { return p == pos; });

            if (pos == this->start) {
                cout << rang::fg::black << rang::bg::magenta << " A " << rang::style::reset;
            } else if (pos == this->goal) {
                cout << rang::fg::black << rang::bg::blue << " B " << rang::style::reset;
            } else if (this->walls[i][j]) {
                cout << "▓▓▓" << rang::style::reset;
            } else if (is_sol_pos) {
                cout << rang::fg::black << rang::bg::green << " • " << rang::style::reset;
            } else if (is_expl_pos) {
                cout << rang::fg::black << rang::bg::yellow << " · " << rang::style::reset;
            } else {
                cout << " · ";
            }
        }
        cout << endl;
    }
}

int Maze::manhattan_distance(Position pos1, Position pos2) {
    int x = abs(get<0>(pos1) - get<0>(pos2));
    int y = abs(get<1>(pos1) - get<1>(pos2));
    return x + y;
}

vector<Position> Maze::possible_moves(Position pos) {
    vector<Position> result;
    int row = get<0>(pos);
    int col = get<1>(pos);

    if (row > 0 && !this->walls[row - 1][col]) {
        result.push_back(make_tuple(row - 1, col));
    }
    if (row < get<0>(this->dimensions) - 1 && !this->walls[row + 1][col]) {
        result.push_back(make_tuple(row + 1, col));
    }
    if (col > 0 && !this->walls[row][col - 1]) {
        result.push_back(make_tuple(row, col - 1));
    }
    if (col < get<1>(this->dimensions) - 1 && !this->walls[row][col + 1]) {
        result.push_back(make_tuple(row, col + 1));
    }

    return result;
}

File: src/maze.py
class Maze:
    def __init__(self, maze_str):
        self.start = None
        self.goal = None
        self.dimensions = None
        self.walls = []
        self.explored_states = []
        self.solution = None

        lines = maze_str.split("\n")
        for i, line in enumerate(lines):
            row = []
            for j, char in enumerate(line):
                if char == "A":
                    self.start = (i, j)
                elif char == "B":
                    self.goal = (i, j)
                row.append(char == "#")
            self.walls.append(row)

        self.dimensions = (len(lines), len(lines[0]))

    def solve(self):
        frontier = [Node(self.start)]

        while frontier:
            if len(frontier) > 1:
                frontier.sort(key=lambda node: self.manhattan_distance(self.goal, node.position))

            node = frontier.pop()

            if node.position == self.goal:
                self.solution = node
                return

            self.explored_states.append(node)

            for pos in self.possible_moves(node.position):
                neighbour = Node(pos, node)
                if not any(n.position == neighbour.position for n in self.explored_states) and \
                   not any(n.position == neighbour.position for n in frontier):
                    frontier.append(neighbour)

    def print_solution(self):
        sol_pos = []
        if self.solution:
            node = self.solution
            while node.parent:
                sol_pos.append(node.position)
                node = node.parent

        expl_pos = [node.position for node in self.explored_states]

        count = 0
        while count < len(expl_pos):
            self.print_maze([], expl_pos[:count])
            Utils.move_up_by(self.dimensions[0])
            count += 1

        if self.solution:
            count = 0
            while count < len(sol_pos):
                self.print_maze(sol_pos[:count], expl_pos)
                Utils.move_up_by(self.dimensions[0])
                count += 1

        self.print_maze(sol_pos, expl_pos)
        print()

        if self.solution:
            print(f"Path length:     {len(sol_pos)}")
        else:
            print("No solution found")
        print(f"Explored states: {len(self.explored_states)}")

    def print_maze(self, solution_pos, explored_pos):
        for i, row in enumerate(self.walls):
            for j, is_wall in enumerate(row):
                pos = (i, j)
                if pos == self.start:
                    print(" A ", end="")
                elif pos == self.goal:
                    print(" B ", end="")
                elif is_wall:
                    print("▓▓▓", end="")
                elif pos in solution_pos:
                    print(" • ", end="")
                elif pos in explored_pos:
                    print(" · ", end="")
                else:
                    print(" · ", end="")
            print()

    @staticmethod
    def manhattan_distance(pos1, pos2):
        return abs(pos1[0] - pos2[0]) + abs(pos1[1] - pos2[1])

    def possible_moves(self, pos):
        result = []
        row, col = pos

        if row > 0 and not self.walls[row - 1][col]:
            result.append((row - 1, col))
        if row < self.dimensions[0] - 1 and not self.walls[row + 1][col]:
            result.append((row + 1, col))
        if col > 0 and not self.walls[row][col - 1]:
            result.append((row, col - 1))
        if col < self.dimensions[1] - 1 and not self.walls[row][col + 1]:
            result.append((row, col + 1))

        return result

class Node:
    def __init__(self, position, parent=None):
        self.position = position
        self.parent = parent
